### PR TITLE
Prevent IllegalArgumentExceptions in BrokenInputStream/Reader/OutputStream/Writer

### PR DESCRIPTION
**Pull request by robtimus**
…to prevent IllegalArgumentExceptions in try-with-resources

### CODE REVIEW

File: src/main/java/org/apache/commons/io/output/BrokenWriter.java [hunk 1]
```diff
@@ -54,7 +55,17 @@ public BrokenWriter() {
...
      */
     public BrokenWriter(final IOException exception) {
-        this.exception = exception;
+        this(() -> exception);
+    }
+
+    /**
+     * Creates a new writer that always throws an {@link IOException}.
+     *
+     * @param exceptionSupplier a supplier for the exception to be thrown
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/io/output/BrokenWriter.java:64]: End sentences in a period.

File: src/test/java/org/apache/commons/io/input/BrokenInputStreamTest.java [hunk 2]
```diff
@@ -67,4 +67,18 @@ public void testSkip() {
         assertEquals(exception, assertThrows(IOException.class, () -> stream.skip(1)));
     }
 
+    @Test
+    public void testTryWithResources() {
+        IOException thrown = assertThrows(IOException.class, () -> {
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/io/input/BrokenInputStreamTest.java:72]: Use final where you can.

File: src/test/java/org/apache/commons/io/input/BrokenReaderTest.java [hunk 3]
```diff
@@ -86,4 +86,19 @@ public void testSkip() {
...
+    @Test
+    public void testTryWithResources() {
+        IOException thrown = assertThrows(IOException.class, () -> {
+            try (Reader newReader = new BrokenReader()) {
+                newReader.read();
+            }
+        });
+        assertEquals("Broken reader", thrown.getMessage());
+
+        Throwable[] suppressed = thrown.getSuppressed();
```
Comment (garydgregory) [inline src/test/java/org/apache/commons/io/input/BrokenReaderTest.java:98]: Use final where you can.

**Review by garydgregory — CHANGES_REQUESTED** [review]
@robtimus 
Looks good mod a few comments I've sprinkled that should apply throughout.
TY!

**Comment by garydgregory** [comment]
https://issues.apache.org/jira/browse/IO-729

**Comment by robtimus** [comment]
@garydgregory I've added periods and made all new variables in the tests final.

**Pull request merged**